app/src/common/shared/com/igalia/wolvic/browser/WebAppsStore.java

```diff
@@ -0,0 +1,131 @@
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+
+import com.google.gson.Gson;
+import com.igalia.wolvic.R;
+import com.igalia.wolvic.ui.adapters.WebApp;
+import com.igalia.wolvic.utils.StringUtils;
+import com.igalia.wolvic.utils.SystemUtils;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+public class WebAppsStore implements SharedPreferences.OnSharedPreferenceChangeListener {
+
+    protected final String LOGTAG = SystemUtils.createLogtag(this.getClass());
+
+    private Context mContext;
+    private List<WebApp> mWebApps;
+    private Set<WebAppsListener> mListeners;
+    private SharedPreferences mPrefs;
+
+    public WebAppsStore(Context context) {
+        mContext = context.getApplicationContext();
+        mWebApps = new ArrayList<>();
+        mListeners = new LinkedHashSet<>();
+
+        mPrefs = PreferenceManager.getDefaultSharedPreferences(mContext);
+        mPrefs.registerOnSharedPreferenceChangeListener(this);
+
+        mWebApps = new ArrayList<>();
+        updateWebAppsListFromStorage();
+    }
+
+    private void updateWebAppsListFromStorage() {
+        String json = SettingsStore.getInstance(mContext).getWebAppsData();
+
+        if (StringUtils.isEmpty(json)) {
+            mWebApps.clear();
+            notifyListeners();
+        } else {
+            try {
+                Gson gson = new Gson();
+                WebApp[] webAppsArray = gson.fromJson(json, WebApp[].class);
+                mWebApps.clear();
+                mWebApps.addAll(Arrays.asList(webAppsArray));
+                notifyListeners();
+            } catch (RuntimeException e) {
+                Log.w(LOGTAG, "retrieveListFromStorage: error parsing stored data: " + e.getMessage());
+
+                // the stored data is invalid, so we need to clear it
+                SettingsStore.getInstance(mContext).setWebAppsData("");
+                mWebApps.clear();
+                notifyListeners();
+            }
+        }
+    }
+
+    private void saveWebAppsListToStorage() {
+        Gson gson = new Gson();
+        WebApp[] webAppsArray = mWebApps.toArray(new WebApp[]{});
+        String json = gson.toJson(webAppsArray, WebApp[].class);
+        SettingsStore.getInstance(mContext).setWebAppsData(json);
+    }
+
+    /**
+     * @return {@code true} if the list did not already contain the specified Web app,
+     * {@code false} if the Web app was already in the list (in which case, it was updated).
+     */
+    public boolean addWebApp(@NonNull WebApp webAppToAdd) {
+        for (WebApp webApp : mWebApps) {
+            if (webApp.getId().equals(webAppToAdd.getId())) {
+                // the Web app is already in the list, we update it
+                webApp.copyFrom(webAppToAdd);
+                notifyListeners();
+                return false;
+            }
+        }
```

Comment:
I guess your assumption is that this is not a frequent operation and also the list of web apps is not really huge, so we can live comfortably with a sequential search. I agree, I don't think we need a hash for this case, but perhaps it deserves a comment.

Reply:
Now that I think about it, since you have to define the `equals()` method in the WebApp perhaps you could simply use here `mWebApps.contains()` instead of the for loop + if.

Reply:
I have replaced the list with a `LinkedHashMap`.